File: src/softwaredesign/src/main/java/com/softwaredesign/softwaredesign/controller/GroupController.java
package com.softwaredesign.softwaredesign.controller;

import com.softwaredesign.softwaredesign.entity.Group;
import com.softwaredesign.softwaredesign.exception.DatabaseAccessException;
import com.softwaredesign.softwaredesign.repository.GroupRepository;
import lombok.AllArgsConstructor;
import org.jetbrains.annotations.NotNull;
import org.springframework.dao.DataAccessException;
import org.springframework.stereotype.Controller;

import java.util.List;

@Controller
@AllArgsConstructor
public class GroupController {
    private GroupRepository groupRepository;

    public void addGroup(String name,
                         Integer numberOfStudents) {
        try {
            Group group = Group.builder()
                    .name(name)
                    .numberOfStudents(numberOfStudents)
                    .build();
            groupRepository.save(group);
        } catch (DataAccessException e) {
            throw new DatabaseAccessException(e.getMessage());
        }

    }

    public void addGroup(Group group) {
        try {
            groupRepository.save(group);
        } catch (DataAccessException e) {
            throw new DatabaseAccessException(e.getMessage());
        }
    }

    public Group getGroup(Long id) {
        Group returnedGroup;
        try {
            returnedGroup = groupRepository.getGroupsByGroupId(id);
        } catch (DataAccessException e) {
            throw new DatabaseAccessException(e.getMessage());
        }
        return returnedGroup;
    }

    public void deleteGroup(Group group) {
        try {
            groupRepository.delete(group);
        } catch (DataAccessException e) {
            throw new DatabaseAccessException(e.getMessage());
        }
    }

    public void deleteGroup(Long id) {
        try {
            groupRepository.deleteById(id);
        } catch (DataAccessException e) {
            throw new DatabaseAccessException(e.getMessage());
        }
    }

    public void updateGroupStudentNumbers(Group group,
                                         Integer numberOfStudents) {
        try {
            group.setNumberOfStudents(numberOfStudents);
            groupRepository.save(group);
        } catch (DataAccessException e) {
            throw new DatabaseAccessException(e.getMessage());
        }
    }

    public List<Group> getAllGroups() {
        List<Group> groups;
        try {
            groups =  groupRepository.findAll();
        } catch (DataAccessException e) {
            throw new DatabaseAccessException(e.getMessage());
        }

        return groups;
    }

    public void updateGroupName(@NotNull Group group,
                                String name) {
        try {
            group.setName(name);
            groupRepository.save(group);
        } catch (DataAccessException e) {
            throw new DatabaseAccessException(e.getMessage());
        }
    }
}

File: cpp_impl/controllers/groupcontroller.cpp
#include "groupcontroller.h"


GroupController::~GroupController()
{

}

void GroupController::addGroup(const std::string name, const int teacherID)
{
    this->groupRepository->addNewGroup(name, teacherID);
}

Group GroupController::getGroup(const std::string &groupName)
{
    return this->groupRepository->getGroup(groupName);
}

std::vector<Group> GroupController::getAllGroups()
{
    return this->groupRepository->getAllGroups();
}

std::vector<Group> GroupController::getAllGroupsByTeacherId(const int id)
{
    return this->groupRepository->getTeacherGroups(id);
}

void GroupController::changeGroupName(const int id, std::string newGroupName)
{
    this->groupRepository->changeGroupName(id, newGroupName);
}

void GroupController::changeGroupTeacherId(const int id, const int teacherId)
{
    this->groupRepository->changeGroupTeacherId(id, teacherId);
}

void GroupController::deleteGroup(const int id)
{
    this->groupRepository->deleteGroup(id);
}

GroupController::GroupController(GroupRepository *groupRepository) : groupRepository(groupRepository)
{}
